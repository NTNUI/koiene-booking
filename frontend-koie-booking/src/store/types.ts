// rootState = generic type, defines the root state type
import AdminBooking, { AdminBookingDictionary } from '@/types/admin/AdminBooking';
import { AdminReportKoieDictionary } from '@/types/admin/AdminReports';

export interface RootState {
  version: string;
  koie: KoieState;
  booking: BookingState;
  auth: AuthState;
  avalanche: AvalancheState;
  adminBookings: AdminBookingsState;
}

export interface KoieState {
  currentLocale: string;
  isLoading: boolean;
  koieData: Koie;
  error: boolean;
  allKoier: [];
}

export interface Koie {
  location: {
    latitude: string;
    longitude: string;
    area: string;
    difficultyInfo_eng: string;
    difficultyInfo_nor: string;
    altitude: number;
    terrain_eng: string;
    terrain_nor: string;
    map_pdf: string;
    kartblad: string;
  };
  description: {
    yr_link_eng: string;
    yr_link_nor: string;
    description_eng: string;
    description_nor: string;
    directions_eng: string;
    directions_nor: string;
    parking_eng: string;
    parking_nor: string;
  };
  beds_available_in_booking_window: Array<BedsAvailable>;
  name: string;
  number_of_beds: number;
  booking_window: number;
  price_member: number;
  price_not_member: number;
}

export interface BedsAvailable {
  date: string;
  bedsAvailable: number;
}

export interface BookingState {
  isLoading: boolean;
  dateFrom: string;
  dateTo: string;
  beds: number;
  step: number;
  guests: Array<Guest>;
  numberOfMembers: number;
  numberOfNonMembers: number;
  edited: boolean;
  validForm: boolean;
  unavailableDates: [];
  availableBeds: number;
  noDatesAvailable: boolean;
  error: boolean;
  bookingData: BookingInfo;
  paymentConfirmed: boolean;
}

export interface ReportState {
  validForm: boolean;
  edited: boolean;
  isLoading: boolean;
  error: boolean;
  step: number;
  reportData: ReportData;
  bookingInfo: BookingInfo;
}

export interface ReportData {
  id: number;
  booking_uuid: string;
  firewood: number;
  chopped_up_wood: number;
  smoke_detector_is_working: boolean;
  gas_is_full: boolean;
  gas_burner_primus: number;
  axe: number;
  hammer: number;
  saw: number;
  saw_blade: number;
  saw_bench: number;
  spade: number;
  kerosene_lamp: number;
  detergent: number;
  dishware: number;
  cookware: number;
  cabin_book: number;
  candle_holders: number;
  fire_blanket: number;
  fire_extinguisher: number;
  other_faults: string;
  boat_status: number;
  canoe_status: number;
  life_jackets_status: number;
  feedback: string;
}

export interface AvalancheState {
  warningLevels: Array<Warning>;
  disableBooking: boolean;
  warningData: Array<WarningResult>;
  koierToCheck: Array<string>;
}

export interface Warning {
  day: string;
  level: number;
}

export interface WarningResult {
  dangerLevel: number;
  validTo: string;
}

export interface CreateBookingInfo {
  koie: string;
  from_date: string;
  to_date: string;
  guests: Array<Guest>;
  guests_member: number;
  guests_not_member: number;
}

export interface BookingInfo {
  id: number;
  uuid: string;
  user: number;
  koie: string;
  price: number;
  booking_transaction_id: string;
  from_date: string;
  to_date: string;
  guests_member: number;
  guests_not_member: number;
}

export interface AuthState {
  isAdmin: boolean;
  tokens: {
    access: string | null;
    refresh: string | null;
  };
}

export interface Guest {
  name: string;
  number: string;
  email: string;
  isMember: boolean;
}

export interface AdminBookingsState {
  cabinsWithBookings: AdminBookingDictionary;
  startDate: string;
}

export interface AdminReportsState {
  chosenCabin: string;
  expanded: boolean;
  cabins: AdminReportKoieDictionary;
}
