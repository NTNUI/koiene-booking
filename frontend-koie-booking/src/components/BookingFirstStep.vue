<template>
  <ErrorCard v-if="apiError" />
  <LoadingSpinner v-else-if="isLoading" />
  <v-layout v-else :class="$style.container" :dark="true">
    <h1 :class="$style.heading">{{ $t('booking.reservation') }} {{ koieTitle }}</h1>
    <v-layout :class="$style.separator">
      <h3>{{ $t('booking.dates') }}</h3>
      <v-layout class="px-1">
        <Calendar :vbeds="false" :row="true" />
      </v-layout>
    </v-layout>
    <v-layout :class="$style.separator">
      <h3>{{ $t('booking.bed_reservation') }}</h3>
      <v-layout :class="$style.availableBeds">
        <v-icon :class="$style.icon" v-text="'mdi-bed'"></v-icon>
        <p :class="$style.availableBedsText">
          {{ bedsAvailable }} / {{ bedsTotal }} {{ $t('booking.available_beds') }}
        </p>
      </v-layout>
      <v-layout :class="$style.chooseBeds">
        <p style="margin-bottom: 0" :class="$style.chooseText">{{ $t('booking.choose_numbers') }}</p>
        <v-text-field
          v-model="bedsChosen"
          :value="bedsChosen"
          :class="$style.bedPicker"
          min="1"
          :max="bedsAvailable"
          :dense="true"
          hide-details
          single-line
          outlined
          type="number"
          @change="(e) => setBeds(e)"
        />
      </v-layout>
      <v-alert
        v-if="error"
        v-model="error"
        transition="expand-transition"
        :dismissible="true"
        :class="$style.errorMsg"
        type="info"
        @click="error = false"
      >
        {{ $t('calendar.errormsg') }} {{ bedsAvailable }}
      </v-alert>
    </v-layout>
  </v-layout>
</template>

<script lang="ts">
import Calendar from '../components/Calendar.vue';
import ErrorCard from '../components/ErrorCard.vue';
import LoadingSpinner from '../components/LoadingSpinner.vue';
import Vue from 'vue';
import { BookingFirstStepData } from '../types/booking';
import BookingSecondStepVue from './BookingSecondStep.vue';

export default Vue.extend({
  name: 'BookingFirstStepVue',
  components: {
    Calendar,
    ErrorCard,
    LoadingSpinner
  },
  data(): BookingFirstStepData {
    return {
      bedsTotal: 8,
      bedsChosen: 1,
      error: false
    };
  },
  computed: {
    koieTitle(): string {
      return this.$store.state.koie.koieData.name || this.$route.params.id;
    },
    apiError(): boolean {
      return this.$store.state.koie.error;
    },
    isLoading(): boolean {
      return this.$store.state.koie.isLoading;
    },
    bedsAvailable(): number {
      return this.$store.state.booking.availableBeds;
    }
  },
  created() {
    this.getTotalBeds();
    if (this.$store.state.booking.beds <= this.bedsAvailable) {
      this.bedsChosen = this.$store.state.booking.beds;
    }
  },
  methods: {
    setBeds(number: number) {
      if (this.$store.state.booking.edited) {
        this.$store.dispatch('booking/SET_EDITED', false);
      }
      if (number <= this.bedsAvailable) {
        this.$store.dispatch('booking/SET_BEDS', number);
      } else {
        this.bedsChosen = 1;
        this.$store.dispatch('booking/SET_BEDS', 1);
        this.error = true;
      }
    },
    goToNextStep() {
      this.$store.dispatch('SET_STEP', 2);
    },
    async getTotalBeds() {
      if (this.$store.state.koie.koieData.name.toLowerCase() !== this.koieTitle.toLowerCase()) {
        await this.$store.dispatch('koie/FETCH_DATA', this.koieTitle.toLowerCase());
      }
      this.bedsTotal = await this.$store.state.koie.koieData.number_of_beds;
    }
  }
});
</script>

<style lang="scss" module>
.container {
  display: flex;
  flex-direction: column;
}
.heading {
  margin-top: 20px;
  align-self: center;
}
.separator {
  border-left: solid 4px;
  border-color: #2f3a50;
  display: flex;
  flex-direction: column;
  margin-top: 24px;
}

.separator > h3 {
  padding: 16px;
}
.icon {
  margin-right: 8px;
  margin-bottom: 6px;
}
.availableBeds {
  display: flex;
  flex-direction: row;
  padding-left: 16px;
  padding-right: 16px;
}
.availableBedsText {
  margin-top: 10px;
}
.chooseBeds {
  display: flex;
  align-items: center;
  padding-left: 16px;
  padding-right: 16px;
}
.chooseText {
  margin-right: 20px;
}
.bedPicker {
  max-width: 60px;
  margin-left: 20px;
}
.errorMsg {
  margin-top: 10px;
}
</style>
