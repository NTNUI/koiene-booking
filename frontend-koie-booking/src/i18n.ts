import Vue from 'vue';
import VueI18n, { LocaleMessages } from 'vue-i18n';
import enJson from './locales/en.json';

Vue.use(VueI18n);

function loadLocaleMessages(): LocaleMessages {
  // Testing currently only supports manually importing of json-files. English is the only language that gets loaded now
  let locales: any;
  const messages: LocaleMessages = {};
  if (process.env.NODE_ENV === 'test') {
    messages['en'] = enJson;
  } else {
    locales = require.context('./locales', true, /[A-Za-z0-9-_,\s]+\.json$/i);
    locales.keys().forEach((key: any) => {
      const matched = key.match(/([A-Za-z0-9-_]+)\./i);
      if (matched && matched.length > 1) {
        const locale = matched[1];
        messages[locale] = locales(key);
      }
    });
  }
  return messages;
}

export default new VueI18n({
  locale: process.env.VUE_APP_I18N_LOCALE || 'en',
  fallbackLocale: process.env.VUE_APP_I18N_FALLBACK_LOCALE || 'en',
  messages: loadLocaleMessages()
});
