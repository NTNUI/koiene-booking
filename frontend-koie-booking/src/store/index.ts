import Vue from 'vue';
import Vuex, { StoreOptions } from 'vuex';
import axios from 'axios';
import VueAxios from 'vue-axios';
import { RootState } from './types';
import { koie } from './modules/users/koie/index';
import { booking } from './modules/users/booking/index';
import { avalanche } from './modules/users/avalanche/index';
import { auth } from './modules/users/auth/index';

Vue.use(Vuex, VueAxios, axios);

const store: StoreOptions<RootState> = {
  modules: {
    koie,
    booking,
    avalanche,
    auth
  }
};

export default new Vuex.Store<RootState>(store);

export { store as storeConfig };
