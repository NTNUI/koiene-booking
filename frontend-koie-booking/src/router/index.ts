import Vue from 'vue';
import VueRouter from 'vue-router';
import store from '@/store';

Vue.use(VueRouter);

const routes = [
  {
    path: '/',
    name: 'AllKoier',
    component: () => import(/* webpackChunkName: "about" */ '../views/AllKoier.vue'),
    meta: { requiresAuth: false, requiresAdmin: false }
  },
  {
    path: '/koie/:id',
    name: 'koie',
    component: () => import('../views/Koie.vue'),
    props: true,
    meta: { requiresAuth: false, requiresAdmin: false }
  },
  {
    path: '/booking/:id',
    name: 'booking',
    component: () => import('../views/Booking.vue'),
    meta: { requiresAuth: true, requiresAdmin: false }
  },
  {
    path: '/login',
    name: 'Login',
    component: () => import('../views/Login.vue'),
    meta: { requiresAuth: false, requiresAdmin: false }
  },
  {
    path: '/admin',
    name: 'Admin',
    component: () => import('../views/Admin.vue'),
    meta: { requiresAuth: true, requiresAdmin: true }
  }
];

const router = new VueRouter({
  routes
});

export function beforeEach(to: any, from: any, next: any) {
  const isLoggedIn = store.getters['auth/isLoggedIn'];
  const isAdmin = store.getters['auth/isAdmin'];
  const requiresAuth = to.matched.some((record: any) => record.meta.requiresAuth);
  const requiresAdmin = to.matched.some((record: any) => record.meta.requiresAdmin);

  if (requiresAdmin && !isAdmin) next({ name: 'Login', query: { redirect: to.path } });
  else if (requiresAuth && !isLoggedIn) next({ name: 'Login', query: { redirect: to.path } });
  else if (to.name === 'Login' && isLoggedIn) next('/');
  else next();
}

router.beforeEach((to, from, next) => beforeEach(to, from, next));

export default router;
export { routes };
