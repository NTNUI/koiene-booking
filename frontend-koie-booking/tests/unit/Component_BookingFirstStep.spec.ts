import Vue from 'vue';
import Vuetify from 'vuetify';
import VueRouter from 'vue-router';
import Vuex from 'vuex';
import { storeConfig } from '@/store';
import { cloneDeep } from 'lodash';
import i18n from '@/i18n';
import scssVars from '@/styles/variables.scss';
import flushPromises from 'flush-promises';
import mockAxios from 'jest-mock-axios';
import { routes } from '@/router';

Vue.use(Vuetify);

// Utilities
import { mount, createLocalVue } from '@vue/test-utils';

// Components or views
import BookingFirstStep from '../../src/components/BookingFirstStep.vue';

describe('Component BookingFirstStep.vue', () => {
  // Router not needed for this test-suite
  let wrapper: any;
  let localVue: any;
  let vuetify: any;
  let store: any;

  beforeEach(() => {
    localVue = createLocalVue();
    vuetify = new Vuetify();
    localVue.use(Vuetify);
    localVue.use(Vuex);
    // Hard resets the store between tests
    store = new Vuex.Store(cloneDeep(storeConfig));
    localVue.prototype.$scssVars = scssVars;
  });

  afterEach(() => {
    mockAxios.reset();
  });

  it('Matches snapshot', async () => {
    wrapper = mount(BookingFirstStep, {
      localVue,
      vuetify,
      i18n,
      store,
      mocks: {
        $route: {
          params: { id: 'flåkoia' }
        }
      }
    });

    const responseObj = { data: { koie: { name: 'flåkoia' } } };
    mockAxios.mockResponse(responseObj);
    await flushPromises();

    expect(wrapper).toMatchSnapshot();
  });
});
