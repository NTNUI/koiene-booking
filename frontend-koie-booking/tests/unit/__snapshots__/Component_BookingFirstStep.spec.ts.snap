// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component BookingFirstStep.vue Matches snapshot 1`] = `
<div class="layout dark container">
  <h1 class="heading">Reservation for flåkoia</h1>
  <div class="layout separator">
    <h3>Dates</h3>
    <div class="layout px-1">
      <div class="container row row">
        <div class="v-input v-input--is-label-active v-input--is-dirty theme--dark v-text-field rowTextFieldArrival">
          <div class="v-input__prepend-outer">
            <div class="v-input__icon v-input__icon--prepend"><i aria-hidden="true" class="v-icon notranslate material-icons theme--dark">event</i></div>
          </div>
          <div class="v-input__control">
            <div class="v-input__slot">
              <div class="v-text-field__slot"><label for="input-12" class="v-label v-label--active theme--dark" style="left: 0px; position: absolute;">Arrival</label><input id="input-12" type="text"></div>
            </div>
            <div class="v-text-field__details">
              <div class="v-messages theme--dark"><span name="message-transition" tag="div" class="v-messages__wrapper"></span></div>
            </div>
          </div>
        </div>
        <div class="v-menu">
          <!---->
        </div>
        <div class="v-menu">
          <!---->
        </div>
        <div class="v-input v-input--is-label-active v-input--is-dirty theme--dark v-text-field rowTextFieldDeparture bookingTabletRemovePadding">
          <div class="v-input__prepend-outer">
            <div class="v-input__icon v-input__icon--prepend"><i aria-hidden="true" class="v-icon notranslate material-icons theme--dark">event</i></div>
          </div>
          <div class="v-input__control">
            <div class="v-input__slot">
              <div class="v-text-field__slot"><label for="input-18" class="v-label v-label--active theme--dark" style="left: 0px; position: absolute;">Departure</label><input id="input-18" type="text"></div>
            </div>
            <div class="v-text-field__details">
              <div class="v-messages theme--dark"><span name="message-transition" tag="div" class="v-messages__wrapper"></span></div>
            </div>
          </div>
        </div>
        <div class="divBeds">
          <!---->
          <!---->
          <!---->
        </div>
        <!---->
      </div>
    </div>
  </div>
  <div class="layout separator">
    <h3>Bed Reservation</h3>
    <div class="layout availableBeds"><i aria-hidden="true" class="v-icon notranslate mdi mdi-bed theme--light icon"></i>
      <p class="availableBedsText">
        Infinity / available beds
      </p>
    </div>
    <div class="layout chooseBeds">
      <p class="chooseText" style="margin-bottom: 0px;">Choose number of beds: </p>
      <div class="v-input v-input--hide-details v-input--is-label-active v-input--is-dirty v-input--dense theme--light v-text-field v-text-field--single-line v-text-field--enclosed v-text-field--outlined bedPicker">
        <div class="v-input__control">
          <div class="v-input__slot">
            <fieldset aria-hidden="true">
              <legend><span>​</span></legend>
            </fieldset>
            <div class="v-text-field__slot"><input min="1" max="Infinity" id="input-24" type="number"></div>
          </div>
        </div>
      </div>
    </div>
    <!---->
  </div>
</div>
`;
